### PR TITLE
 Add cleanup possibility on processor startup

### PR DESCRIPTION
**Pull request by fapaul**

### CODE REVIEW

File: src/main/java/com/bakdata/common_kafka_streams/KafkaStreamsApplication.java [hunk 1]
```diff
@@ -101,28 +106,44 @@ protected static void startApplication(final KafkaStreamsApplication app, final
...
         final var kafkaProperties = this.getKafkaProperties();
-        this.streams = new KafkaStreams(this.createTopology(), kafkaProperties);
 
         if (this.forceReprocessing) {
+            if (!this.inputTopic.isBlank()) {
+                this.runResetter(this.inputTopic, this.brokers, this.getUniqueAppId());
+            }
             this.streams.cleanUp();
+            try {
+                Thread.sleep(5000);
```
Comment (lawben) [inline src/main/java/com/bakdata/common_kafka_streams/KafkaStreamsApplication.java]:
```suggestion
                // Wait here to let cleanUp and resetting finish .
                Thread.sleep(5000);
```

File: src/test/java/com/bakdata/common_kafka_streams/test_applications/Mirror.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.streams.StreamsBuilder;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.kstream.KStream;
+import picocli.CommandLine.Command;
+
+@Command(name = "mirror")
```
Comment (SvenLehmann) [inline src/test/java/com/bakdata/common_kafka_streams/test_applications/Mirror.java]: Should not be necessary here

File: src/main/java/com/bakdata/common_kafka_streams/KafkaStreamsApplication.java [hunk 3]
```diff
@@ -101,28 +106,45 @@ protected static void startApplication(final KafkaStreamsApplication app, final
...
         if (this.forceReprocessing) {
+            if (!this.inputTopic.isBlank()) {
+                this.runResetter(this.inputTopic, this.brokers, this.getUniqueAppId());
+            }
             this.streams.cleanUp();
+            try {
+                // Wait here to let cleanUp and resetting finish .
+                Thread.sleep(5000);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
```
Comment (lawben) [inline src/main/java/com/bakdata/common_kafka_streams/KafkaStreamsApplication.java]: This should probably propagate and end the application.

File: src/test/java/com/bakdata/common_kafka_streams/integration/ReprocessingTest.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    private List<KeyValue<String, String>> readFromTopic(String topic) throws InterruptedException {
+        final ReadKeyValues<String, String> readRequest = ReadKeyValues.from(topic).useDefaults();
+        return this.kafkaCluster.read(readRequest);
+    }
+
+    private void runAndAssert(final int expectedMessageCount) throws InterruptedException {
+        this.mirror.run();
+        Thread.sleep(5000);
```
Comment (SvenLehmann) [inline src/test/java/com/bakdata/common_kafka_streams/integration/ReprocessingTest.java]: Comment?

File: src/test/java/com/bakdata/common_kafka_streams/integration/ReprocessingTest.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        final SendValuesTransactional<String> sendRequest =
+                SendValuesTransactional.inTransaction(this.mirror.getInputTopic(),
+                        Arrays.asList("a", "b", "c")).useDefaults();
+        this.kafkaCluster.send(sendRequest);
+
+        this.runAndAssert(3);
+
+        this.runAndAssert(3);
+
+        Thread.sleep(10000);
```
Comment (SvenLehmann) [inline src/test/java/com/bakdata/common_kafka_streams/integration/ReprocessingTest.java]: Comment?

File: src/test/java/com/bakdata/common_kafka_streams/WordCountTest.java [hunk 6]
```diff
@@ -37,12 +37,12 @@
...
 
     @RegisterExtension
-    final TestTopology<Object, String> testTopology = new TestTopology<>(this.app::createTopology,
+    final TestTopologyExtension<Object, String> testTopology = new TestTopologyExtension<>(this.app::createTopology,
             this.app.getKafkaProperties());
 
     @Test
     void shouldAggregateSameWordStream() {
-        this.testTopology.input().add("bla")
+        this.testTopology.input(this.app.getInputTopic()).add("bla")
```
Comment (lawben) [inline src/test/java/com/bakdata/common_kafka_streams/WordCountTest.java]:
```suggestion
        this.testTopology.input().add("bla")
```

File: src/test/java/com/bakdata/common_kafka_streams/integration/ReprocessingTest.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        this.mirror = new Mirror();
+        this.mirror.setSchemaRegistryUrl(this.schemaRegistryMockExtension.getUrl());
+        final String inputTopicName = "input";
+        this.mirror.setInputTopic(inputTopicName);
+        final String outputTopicName = "output";
+        this.mirror.setOutputTopic(outputTopicName);
+        this.mirror.setBrokers(this.kafkaCluster.getBrokerList());
+        this.mirror.setProductive(false);
+
+        this.kafkaCluster
```
Comment (lawben) [inline src/test/java/com/bakdata/common_kafka_streams/integration/ReprocessingTest.java]: this looks bad

File: src/test/java/com/bakdata/common_kafka_streams/test_applications/Mirror.java (hunk 2 above)
Comment (lawben) [inline src/test/java/com/bakdata/common_kafka_streams/test_applications/Mirror.java]: do we need this?

File: src/test/java/com/bakdata/common_kafka_streams/test_applications/Mirror.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+@Getter
```
Comment (lawben) [inline src/test/java/com/bakdata/common_kafka_streams/test_applications/Mirror.java]:
```suggestion

```

File: src/test/java/com/bakdata/common_kafka_streams/test_applications/Mirror.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+@Setter
```
Comment (lawben) [inline src/test/java/com/bakdata/common_kafka_streams/test_applications/Mirror.java]:
```suggestion
```

File: src/test/java/com/bakdata/common_kafka_streams/test_applications/WordCount.java [hunk 10]
```diff
@@ -35,24 +38,13 @@
...
 import org.apache.kafka.streams.kstream.KTable;
 import org.apache.kafka.streams.kstream.Produced;
-import picocli.CommandLine;
 import picocli.CommandLine.Command;
 
-import java.util.Arrays;
-import java.util.Properties;
-import java.util.regex.Pattern;
-
 @Command(name = "wordcount")
```
Comment (lawben) [inline src/test/java/com/bakdata/common_kafka_streams/test_applications/WordCount.java]:
```suggestion
```

File: src/test/java/com/bakdata/common_kafka_streams/test_applications/WordCount.java [hunk 11]
```diff
@@ -35,24 +38,13 @@
...
 @Getter
```
Comment (lawben) [inline src/test/java/com/bakdata/common_kafka_streams/test_applications/WordCount.java]:
```suggestion
```

File: src/test/java/com/bakdata/common_kafka_streams/test_applications/WordCount.java [hunk 12]
```diff
@@ -35,24 +38,13 @@
...
 @Setter
```
Comment (lawben) [inline src/test/java/com/bakdata/common_kafka_streams/test_applications/WordCount.java]:
```suggestion
```

**Pull request merged**